### PR TITLE
Adapted lab/component navigation over recommended and "thorough" paths

### PR DESCRIPTION
**Pull request by WeeknightMVP**
Here's my first pass at a PR for #143.  It (manually) introduces rudimentary navigation markup for traversing the recommended path, and branching from/to the recommended path to/from optional labs, which each have an entry point and their own recommended path.

This could be polished, of course.  Right now the navigation sections look straight out of an early '90s text browser...

### CODE REVIEW

**Comment by weaversa** [comment]
I think the answers files shouldn’t link to each other. Consider, instead CRC <-> CRCAnswers. Now it’s easier to peek at the answers and go back to where you were to continue the exercises.

File: README.md [hunk 1]
```diff
@@ -93,6 +93,8 @@ Cryptol](INSTALL.md).
 4. [Language Basics](labs/Language/Basics.md): A resource
    for (most) of the language features you'll need to be successful
    here.
+    * [Style Guide](cryptol-style.md): Cryptol style guide we
```
Comment (weaversa) [inline README.md:96]: That’s a good idea

**Comment by weaversa** [comment]
> I think the answers files shouldn’t link to each other. Consider, instead CRC <-> CRCAnswers. Now it’s easier to peek at the answers and go back to where you were to continue the exercises.

Oh, I see that you did both. I guess that’s ok

**Comment by WeeknightMVP** [comment]
The `Answers` mirroring gets wonky when related exercises don't have them. Perhaps `Answers` should just link back to the associated `Exercise`, or to the same related (non-`Answers`) links.

**Comment by WeeknightMVP** [comment]
There's a new Python script (`l4y3rc4k3.py`) that generates (from `deps.yml` and `path.gv.jinja`) graphical views of the overall course (linked from `README.md`) and optional modules with multiple components (linked from respective Markdown for Cryptol and SAW Demos and Transposition).  The next challenge will be to generate nav links in a similar manner.

**Comment by WeeknightMVP** [comment]
`Answers` navlinks (still manually generated) now point to the same links as respective exercises (except for the links to/from `Answers`).  Now to automate this...

**Comment by weaversa** [comment]
What are you thinking with regard to automation? Perhaps the simplest is a script that tells you what to put at the bottom based on the graph.

File: scripts/l4y3rc4k3/path.gv.jinja [hunk 2]
```diff
@@ -0,0 +1,51 @@
+digraph {{ branch_id }} {
+  //splines = ortho;
+  //concentrate = true;
+
+  rankdir="LR";
+
+  node [shape="box", style="rounded,filled", fillcolor="red", penwidth = 2];
```
Comment (weaversa) [inline scripts/l4y3rc4k3/path.gv.jinja:7]: I think fillcolor="white" would be better

Comment (WeeknightMVP) [reply scripts/l4y3rc4k3/path.gv.jinja:7]: This was just to further distinguish recommended labs from optional ones.  If you would rather not, `fillcolor` settings can be changed or removed in `path.gv.jinja`, the template file from which the Python script generates graphs for the course and optional labs with multiple components.

Comment (weaversa) [reply scripts/l4y3rc4k3/path.gv.jinja]: How does one run the script?

**Comment by WeeknightMVP** [comment]
@weaversa The script currently renders graphs of the suggested path and branches for the course and each multi-component optional lab, based on a template (`path.gv.jinja`) and metadata (`deps.yml`).  Similarly, each Markdown file could be rendered from a simple template with a variable (e.g. `{{ navlinks }}`) that the Python script could render based on the corresponding entry in `deps.yml`.  These Markdown templates could be stored in a directory alongside the Python script (e.g. `scripts/l4y3rc4k3/templates/README.md.jinja`, `scripts/l4y3rc4k3/templates/labs/CRC/CRCAnswers.md.jinja`, etc.) that mirrors the top level.

Likewise, solutions and exercises for each applicable lab could be based on a single template with each placeholder/hint and solution (DRY!), but that's for later and would require some tweaking for consistent presentation among labs.

**Comment by weaversa** [comment]
> @weaversa The script currently renders graphs of the suggested path and branches for the course and each multi-component optional lab, based on a template (`path.gv.jinja`) and metadata (`deps.yml`). Similarly, each Markdown file could be rendered from a simple template with a variable (e.g. `{{ navlinks }}`) that the Python script could render based on the corresponding entry in `deps.yml`. These Markdown templates could be stored in a directory alongside the Python script (e.g. `scripts/l4y3rc4k3/templates/README.md.jinja`, `scripts/l4y3rc4k3/templates/labs/CRC/CRCAnswers.md.jinja`, etc.) that mirrors the top level.
> 
> Likewise, solutions and exercises for each applicable lab could be based on a single template with each placeholder/hint and solution (DRY!), but that's for later and would require some tweaking for consistent presentation among labs.

I see. How can I run the script to render the graphs?

File: scripts/l4y3rc4k3/path.gv.jinja (hunk 2 above)
Comment (WeeknightMVP) [reply scripts/l4y3rc4k3/path.gv.jinja]: Right now the script just generates the graphs in `/misc`. I haven't tried it on any other workspace, but you should be able to do something like:

```
> python3 -m pip install poetry
> cd cryptol-course/scripts/l4y3rc4k3
> poetry install
> poetry shell
> python3 l4y3rc4k3/l4y3rc4k3.py
```

I'll add instructions for this and other scripts when they're more stable.  (And yes, rename `l4y3rc4k3` to something less silly.)

Comment (weaversa) [reply scripts/l4y3rc4k3/path.gv.jinja]: Works!

That's very nice so far.

**Comment by WeeknightMVP** [comment]
The SVGs were misbehaving, so I changed the Markdown to link to them via relative URLs, and changed the metadata to use `.html` extensions rather than `.md`.  This seems to have done the trick for coherent GitHub Pages, but at what price to local browsing...?

**Comment by weaversa** [comment]
I don’t think the sgv was clickable in VSCode. So it’s probably not a big deal. Besides, the layercake works everywhere, right! That’s great!

**Comment by WeeknightMVP** [comment]
I don't see any more broken navlinks or SVGs, but perhaps we should run a GitHub Action to check for broken links.  There are plenty to choose from.

**Review by weaversa — APPROVED** [review]
One thing left is to add IntroTypeHackery into the layercake, then I'm content to merge this.

**Comment by weaversa** [comment]
> Alright. I think I fixed all the broken links I know how to fix. I do not know how to get Github Pages/Jekyll to include a directory of non-html files. I also don't know why the sgv file in README is linking funny. It clicks just fine in the html, but the broken-links action says the file is one directory to the left of where it should be.

Link to the output here: https://github.com/weaversa/cryptol-course/runs/1652790562?check_suite_focus=true

**Comment by WeeknightMVP** [comment]
The `Type Hackery` branch didn't specify where it falls in the course map, so I just branched from `Language Basics`.

**Pull request merged**